src/repr/src/relation.rs

```diff
@@ -657,13 +819,285 @@ impl fmt::Display for NotNullViolation {
+        mut self,
+        name: N,
+        ty: ColumnType,
+    ) -> RelationDescBuilder {
+        let name = name.into();
+        self.columns.push((name, ty));
+        self
+    }
+
+    /// Appends the provided columns to the builder.
+    pub fn with_columns<I, T, N>(mut self, iter: I) -> Self
+    where
+        I: IntoIterator<Item = (N, T)>,
+        T: Into<ColumnType>,
+        N: Into<ColumnName>,
+    {
+        self.columns
+            .extend(iter.into_iter().map(|(name, ty)| (name.into(), ty.into())));
+        self
+    }
+
+    /// Adds a new key for the relation.
+    pub fn with_key(mut self, mut indices: Vec<usize>) -> RelationDescBuilder {
+        indices.sort_unstable();
+        if !self.keys.contains(&indices) {
+            self.keys.push(indices);
+        }
+        self
+    }
+
+    /// Removes all previously inserted keys.
+    pub fn without_keys(mut self) -> RelationDescBuilder {
+        self.keys.clear();
+        assert_eq!(self.keys.len(), 0);
+        self
+    }
+
+    /// Concatenates a [`RelationDescBuilder`] onto the end of this [`RelationDescBuilder`].
+    pub fn concat(mut self, other: Self) -> Self {
+        let self_len = self.columns.len();
+
+        self.columns.extend(other.columns);
+        for k in other.keys {
+            let k = k.into_iter().map(|idx| idx + self_len).collect();
+            self = self.with_key(k);
+        }
+
+        self
+    }
+
+    /// Finish the builder, returning a [`RelationDesc`].
+    pub fn finish(self) -> RelationDesc {
+        let mut desc = RelationDesc::from_names_and_types(self.columns);
+        desc.typ = desc.typ.with_keys(self.keys);
+        desc
+    }
+}
+
+/// Describes a [`RelationDesc`] at a specific version of a [`VersionedRelationDesc`].
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+pub enum RelationVersionSelector {
+    Specific(RelationVersion),
+    Latest,
+}
+
+impl RelationVersionSelector {
+    pub fn specific(version: u64) -> Self {
+        RelationVersionSelector::Specific(RelationVersion(version))
+    }
+}
+
+/// A wrapper around [`RelationDesc`] that provides an interface for adding
+/// columns and generating new versions.
+///
+/// TODO(parkmycar): Using an immutable data structure for RelationDesc would
+/// be great.
+#[derive(Debug, Clone, Serialize)]
+pub struct VersionedRelationDesc {
+    inner: RelationDesc,
+}
+
+impl VersionedRelationDesc {
+    pub fn new(inner: RelationDesc) -> Self {
+        VersionedRelationDesc { inner }
+    }
+
+    /// Adds a new column to this [`RelationDesc`], creating a new version of the [`RelationDesc`].
+    ///
+    /// Note: For building a [`RelationDesc`] see [`RelationDescBuilder::with_column`].
+    #[must_use]
+    pub fn add_column<N, T>(&mut self, name: N, typ: T) -> RelationVersion
+    where
+        N: Into<ColumnName>,
+        T: Into<ColumnType>,
+    {
+        let latest_version = self.latest_version();
```

Comment:
This makes the cost linear in the number of existing fields. (So adding N fields is quadratic.) Not sure if we care?

Reply:
Yeah we could cache the latest version, but this should only get called when a user runs `ALTER TABLE ...` so I'm not concerned about the runtime here